common/scala/src/main/scala/org/apache/openwhisk/core/database/elasticsearch/ElasticSearchActivationStore.scala

```diff
@@ -0,0 +1,467 @@
+  useBatching: Boolean = false)(implicit actorSystem: ActorSystem,
+                                actorMaterializer: ActorMaterializer,
+                                logging: Logging)
+    extends ActivationStore {
+
+  import com.sksamuel.elastic4s.http.ElasticDsl._
+
+  private implicit val executionContextExecutor: ExecutionContextExecutor = actorSystem.dispatcher
+
+  private val httpClientCallback = new HttpClientConfigCallback {
+    override def customizeHttpClient(httpClientBuilder: HttpAsyncClientBuilder): HttpAsyncClientBuilder = {
+      val provider = new BasicCredentialsProvider
+      provider.setCredentials(
+        AuthScope.ANY,
+        new UsernamePasswordCredentials(elasticSearchConfig.username, elasticSearchConfig.password))
+      httpClientBuilder.setDefaultCredentialsProvider(provider)
+    }
+  }
+
+  private val client =
+    ElasticClient(
+      ElasticProperties(s"${elasticSearchConfig.protocol}://${elasticSearchConfig.hosts}"),
+      NoOpRequestConfigCallback,
+      httpClientCallback)
+
+  private val esType = "_doc"
+  private val maxOpenDbRequests = actorSystem.settings.config
+    .getInt("akka.http.host-connection-pool.max-connections") / 2
+  private val batcher: Batcher[IndexRequest, Either[ArtifactStoreException, DocInfo]] =
+    new Batcher(500, maxOpenDbRequests)(doStore(_)(TransactionId.dbBatcher))
+
+  private val minStart = 0L
+  private val maxStart = Instant.now.toEpochMilli + TimeUnit.DAYS.toMillis(365 * 100) //100 years from now
+
+  override def store(activation: WhiskActivation, context: UserContext)(
+    implicit transid: TransactionId,
+    notifier: Option[CacheChangeNotification]): Future[DocInfo] = {
+
+    val start =
+      transid.started(this, LoggingMarkers.DATABASE_SAVE, s"[PUT] 'activations' document: '${activation.docid}'")
+
+    val path = activation.annotations
+      .getAs[String](WhiskActivation.pathAnnotation)
+      .getOrElse(s"${activation.namespace}/${activation.name}")
+    // Escape `_id` field as it's not permitted in ElasticSearch, add `path` field for search, and
+    // convert annotations to JsObject as ElasticSearch doesn't support array with mixed types
+    // response.result can be any type ElasticSearch also doesn't support that, so convert it to a string
+    val response = JsObject(
+      activation.response.toJsonObject.fields
+        .updated("result", JsString(activation.response.result.toJson.compactPrint)))
+    val payload = JsObject(
+      activation.toDocumentRecord.fields - "_id" ++ Map(
+        "path" -> JsString(path),
+        "@timestamp" -> JsString(activation.start.toString),
+        "annotations" -> activation.annotations.toJsObject,
+        "response" -> response))
+
+    val index = generateIndex(activation.namespace.namespace)
+    val op = indexInto(index, esType).doc(payload.toString).id(activation.docid.asString)
+
+    val res = if (useBatching) {
```

Comment:
As `ActivationStore` is always called with `batching` as `true` may be drop the else part or throw exception? Per [coverage stats][1] else part is not covered

[1]: https://codecov.io/gh/apache/openwhisk/pull/4724/diff?src=pr&el=tree-more#D26-127

Reply:
removed `else` part